scripts/CEE/cordon-drain-reboot-worker-node/script.sh

```diff
@@ -0,0 +1,186 @@
+#!/bin/bash
+#
+# Drain and Reboot a worker node.
+# Input:
+#  (Env) WORKER
+#    name of the worker node
+#  (Env) ACTION
+#    action to execute on the worker node
+#  (Env) DRAINMODE
+#    drain options to execute on the worker node
+
+set -e
+set -o nounset
+set -o pipefail
+
+CURRENTDATE=$(date +"%Y-%m-%d %T")
+
+start_job(){
+    echo "Job started at $CURRENTDATE"
+    echo ".................................."
+    echo
+}
+
+finish_job(){
+    echo
+    echo ".................................."
+    echo "Job finished at $CURRENTDATE"
+}
+
+## Function which checks if the node is a Worker node
+check_worker(){
+
+    echo "Checking if ${WORKER} is a worker node..."
+    
+    if (oc get nodes -l node-role.kubernetes.io/worker=,node-role.kubernetes.io/infra!= -oname | grep "${WORKER}") &> /dev/null; then
+        echo "[OK] ${WORKER} is a worker node."
+        echo
+    else
+        echo "[Error] ${WORKER} is not a worker node. Please ensure you are providing the correct name."
+        exit 1
+    fi
+
+}
+
+## Function which check if the node is already cordoning
+check_cordon(){
+    if [[ $(oc get node "${WORKER}" -o jsonpath='{.spec.taints[?(@.key == "node.kubernetes.io/unschedulable")]}') ]]; then
+        echo "[Error] Node is already cordoned"
+        exit 1
+    fi
+}
+
+## Function which check if the node is already uncordoning
+check_uncordon(){
+    if ! [[ $(oc get node "${WORKER}" -o jsonpath='{.spec.taints[?(@.key == "node.kubernetes.io/unschedulable")]}') ]]; then
+        echo "[Error] Node is already uncordoned"
+        exit 1
+    fi
+}
+
+## Function which cordon the worker node
+cordon_worker(){
+    if oc adm cordon "${WORKER}"; then
+        echo "[OK] Node ${WORKER} cordoned successfully"
+        exit 0
+    else
+        echo "[Error] Something went wrong"
+        exit 1
+    fi
+}
+
+## Function which uncordon the worker node
+uncordon_worker(){
+    if oc adm uncordon "${WORKER}"; then
+        echo "[OK] Node ${WORKER} uncordoned successfully"
+        exit 0
+    else
+        echo "[Error] Something went wrong"
+        exit 1
+    fi
+}
+
+## Function which drains the node
+drain_worker(){
+    
+    DRAIN_CMD="oc adm drain ${WORKER}"
+
+    if [ -z ${1} ]; then
+        if ! declare -p DRAINMODE &>/dev/null || [ -z "${DRAINMODE}" ]; then
+            echo "Draining node ${WORKER} with no options."
+            echo
+            ${DRAIN_CMD}
+        else 
+            echo "Draining node ${WORKER} with drain mode '${DRAINMODE}'"
+            echo
+            ${DRAIN_CMD} ${DRAINMODE}
+        fi
+    else
+        echo "Draining node ${WORKER} for rebooting."
+        echo
+        ${DRAIN_CMD} ${1}
```

Comment:
This may fail shellcheck, maybe we can make the options fixed.
If mode==force then execute xxx
else execute yyy.

Reply:
I don't want to make it fixed because drain has a lot of options we can use. So passing it as a variable would helps to choose whatever is needed. I tested and it ran ok.

Reply:
@faldanarh The shell check is failing here. If don't want to change it, could you please add a double quote to the variables here?

Reply:
Hi @faldanarh , I got your point that we want to make the drain command flexible. The `shellcheck` force you to add a double quote for a reason. With a double quote, it will take the whole string as one single argument, so that it can avoid users hijacking the command. Eg, the user can pass "; rm -rf /;" as arguments.